### PR TITLE
Fluid Flow API

### PR DESCRIPTION
**Pull request by OroArmor**
Adds an event when a fluid flows. Yes i know its in the object builder API, I'm waiting for #1136 to be merged, or if we merge these into one larger fluid api

### CODE REVIEW

**Comment by PepperCode1** [comment]
Does this API work for specifying interactions by fluid tag instead of block? Also, does it work for fluid blocks that do not extend `FluidBlock`?

**Comment by OroArmor** [comment]
No, it doesnt work by tag, as each fluid tag should correspond to a block.
Also no, as it hooks directly into the fluid flow method of fluid block. I don't see why someone would make a fluid that doesn't extend fluid block

**Comment by OroArmor** [comment]
OHH i see what you mean by tag, but with the current implementation, i don't think its easily addible

File: fabric-object-builder-api-v1/src/main/java/net/fabricmc/fabric/api/event/fluid/FluidFlowEvents.java [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+						}
+					}
+				}
+			}
+		}
+
+		return null;
+	}
+
+	public interface FluidFlowInteractionEvent {
```
Comment (Juuxel) [inline fabric-object-builder-api-v1/src/main/java/net/fabricmc/fabric/api/event/fluid/FluidFlowEvents.java:111]: Should be named `FluidFlowInteractionCallback` to match other callbacks.

File: fabric-object-builder-api-v1/src/main/java/net/fabricmc/fabric/mixin/object/builder/FluidBlockMixin.java [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+@Mixin(FluidBlock.class)
+public class FluidBlockMixin extends Block {
+	public FluidBlockMixin(Settings settings) {
+		super(settings);
+	}
+
+	@Inject(method = "receiveNeighborFluids", at = @At("HEAD"), cancellable = true)
+	private void receiveNeighborFluids(World world, BlockPos pos, BlockState state, CallbackInfoReturnable<Boolean> cir) {
+		for (Direction direction : Direction.values()) {
+			FluidFlowEvents.FluidFlowInteractionEvent event = FluidFlowEvents.getEvent(this, world.getBlockState(pos.offset(direction)).getBlock(), direction);
```
Comment (Juuxel) [inline fabric-object-builder-api-v1/src/main/java/net/fabricmc/fabric/mixin/object/builder/FluidBlockMixin.java]: Might be a nice micro-optimisation and code clarification to cache `pos.offset(direction)` in a variable (and the block state too).

File: fabric-object-builder-api-v1/src/main/java/net/fabricmc/fabric/api/event/fluid/FluidFlowEvents.java [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+
+	/**
+	 * Gets the event from the following blocks and direction.
+	 *
+	 * @param flowingBlock         The fluid block that flowed.
+	 * @param interactionBlock     The block it interacts with.
+	 * @param interactionDirection The interaction direction
+	 * @return An event if the conditions are met, otherwise {@code null}
+	 */
+	public static @Nullable FluidFlowInteractionEvent getEvent(Block flowingBlock, Block interactionBlock, Direction interactionDirection) {
```
Comment (Juuxel) [inline fabric-object-builder-api-v1/src/main/java/net/fabricmc/fabric/api/event/fluid/FluidFlowEvents.java]: This is a bit concerning, IMO this should allow `Event`-like chaining of callbacks.

File: fabric-object-builder-api-v1/src/testmod/java/net/fabricmc/fabric/test/object/builder/FluidFlowEventTests.java [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+import net.minecraft.sound.SoundEvents;
+import net.minecraft.util.math.Direction;
+
+import net.fabricmc.api.ModInitializer;
+import net.fabricmc.fabric.api.event.fluid.FluidFlowEvents;
+
+public class FluidFlowEventTests implements ModInitializer {
+	@Override
+	public void onInitialize() {
+		FluidFlowEvents.register(Blocks.WATER, Blocks.BLUE_ICE, new Direction[]{Direction.DOWN}, (blockState, blockState2, blockPos, world) -> {
```
Comment (Juuxel) [inline fabric-object-builder-api-v1/src/testmod/java/net/fabricmc/fabric/test/object/builder/FluidFlowEventTests.java]: This should use the parameter names from the functional interface.

**Review by Juuxel — CHANGES_REQUESTED** [review]
Some general comments about this PR:

- ~~The module for this API is wrong, it definitely does not belong in object builders.~~ nevermind, didn't read the PR desc carefully enough
- Using a `Map<Pair<Block, Block>, Event<...>>` could be a better solution: you would get event chaining for free as well as a more usual API. (I'd also change `register` to `Event<...> flowInteraction(Block, Block)` to match normal events)

**Comment by OroArmor** [comment]
The reason i didnt use the current event system is that those are array-based, and this cant easily be array based unless i expose all the data, which would require each event to check if there conditions are met. this is simpler in the sense that you specify the interacting fluid and block to make your callback for

**Comment by Juuxel** [comment]
You can have a map/collection of events matching to each condition, though. That doesn't require any checks in the callbacks.

**Comment by OroArmor** [comment]
Ooooohh yeah i could

**Comment by OroArmor** [comment]
> Using a Map<Pair<Block, Block>, Event<...>> could be a better solution:

I cant use a pair for the block/block matching as the Minecraft pair doesn't have an equals method

**Comment by PepperCode1** [comment]
Personally, I think this API should be a callback that is not attached to anything else so that the user has maximum control and so that if a mod needs multiple interactions they can be put into the same callback for more efficiency.

**Comment by Juuxel** [comment]
@OroArmor You *can* use a `Pair` from some other library or just a custom data class for this as it's an impl detail. I'd still prefer having a public API method returning an `Event` for a block+block+direction combination (or a general event like @PepperCode1 suggested), as that matches the style done everywhere else in FAPI.

**Comment by OroArmor** [comment]
I dont see it hitting multiple callbacks very often, but in the case it does, its a nice to have

**Comment by OroArmor** [comment]
Also i should make an impl class for this as there are way to many impl details in the API part

File: fabric-object-builder-api-v1/src/main/java/net/fabricmc/fabric/api/event/fluid/FluidFlowEvents.java [hunk 5]
```diff
@@ -0,0 +1,123 @@
...
+	 * Registers a new event on a fluid flow. The same two blocks can register a different event for different directions, but the same directions will run the event that was registered first.
+	 *
+	 * @param flowingBlock          The fluid block that flowed.
+	 * @param interactionBlock      The block in one of the {@code interactionDirections}.
+	 * @param interactionDirections The direction to search for {@code interactionBlock}.
+	 * @param interactionEvent      The event to run when the conditions are met.
+	 */
+	public static void register(Block flowingBlock, Block interactionBlock, Direction[] interactionDirections, FluidFlowInteractionEvent interactionEvent) {
+		Map<Block, List<Pair<Direction, Event<FluidFlowInteractionEvent>>>> flowBlockEvents = EVENT_MAP.getOrDefault(flowingBlock, new HashMap<>());
+		List<Pair<Direction, Event<FluidFlowInteractionEvent>>> interactionEvents = flowBlockEvents.getOrDefault(interactionBlock, new ArrayList<>());
```
Comment (liach) [inline fabric-object-builder-api-v1/src/main/java/net/fabricmc/fabric/api/event/fluid/FluidFlowEvents.java:52]: You need to learn to use `computeIfAbsent`

File: fabric-object-builder-api-v1/src/main/java/net/fabricmc/fabric/api/event/fluid/FluidFlowEvents.java [hunk 6]
```diff
@@ -0,0 +1,123 @@
...
+				}
+			}
+		}
+
+		if (!flowBlockEvents.containsKey(interactionBlock)) {
+			flowBlockEvents.put(interactionBlock, interactionEvents);
+		}
+
+		if (!EVENT_MAP.containsKey(flowingBlock)) {
+			EVENT_MAP.put(flowingBlock, flowBlockEvents);
```
Comment (liach) [inline fabric-object-builder-api-v1/src/main/java/net/fabricmc/fabric/api/event/fluid/FluidFlowEvents.java:81]: Had you used computeIfAbsent, you won't need these in the end

File: fabric-object-builder-api-v1/src/main/java/net/fabricmc/fabric/api/event/fluid/FluidFlowEvents.java [hunk 7]
```diff
@@ -0,0 +1,123 @@
...
+	/**
+	 * Gets the event from the following blocks and direction.
+	 *
+	 * @param flowingBlock         The fluid block that flowed.
+	 * @param interactionBlock     The block it interacts with.
+	 * @param interactionDirection The interaction direction
+	 * @return An event if the conditions are met, otherwise {@code null}
+	 */
+	public static @Nullable Event<FluidFlowInteractionEvent> getEvent(Block flowingBlock, Block interactionBlock, Direction interactionDirection) {
+		if (EVENT_MAP.containsKey(flowingBlock)) {
```
Comment (liach) [inline fabric-object-builder-api-v1/src/main/java/net/fabricmc/fabric/api/event/fluid/FluidFlowEvents.java:94]: just check if the get result is null. extra contains check offers no real benefits

**Pull request closed without merging**